src/pybind/mgr/devicehealth/module.py

```diff
@@ -0,0 +1,257 @@
+                continue
+            for device, raw_data in raw_smart_data.items():
+                if device in did_device:
+                    self.log.debug('skipping duplicate %s' % device)
+                    continue
+                did_device[device] = 1
+                data = self.extract_smart_features(raw_data)
+                self.put_device_metrics(ioctx, device, data)
+
+        ioctx.close()
+        return (0, "", "")
+
+    def scrape_device(self, devid):
+        r = self.get("device " + devid)
+        if not r or 'device' not in r.keys():
+            return (-errno.ENOENT, '', 'device ' + devid + ' not found')
+        daemons = r['device'].get('daemons', [])
+        osds = [int(r[4:]) for r in daemons if r.startswith('osd.')]
+        if not osds:
+            return (-errno.EAGAIN, '',
+                    'device ' + devid + ' not claimed by any active OSD daemons')
+        osd_id = osds[0]
+        ioctx = self.open_connection()
+        raw_smart_data = self.do_scrape_osd(osd_id, ioctx, devid=devid)
+        if raw_smart_data:
+            for device, raw_data in raw_smart_data.items():
+                data = self.extract_smart_features(raw_data)
+                self.put_device_metrics(ioctx, device, data)
+        ioctx.close()
+        return (0, "", "")
+
+    def do_scrape_osd(self, osd_id, ioctx, devid=''):
+        self.log.debug('do_scrape_osd osd.%d' % osd_id)
+
+        # scrape from osd
+        result = CommandResult('')
+        self.send_command(result, 'osd', str(osd_id), json.dumps({
+            'prefix': 'smart',
+            'format': 'json',
+            'devid': devid,
+        }), '')
+        r, outb, outs = result.wait()
+
+        try:
+            return json.loads(outb)
+        except:
+            self.log.debug('Fail to parse JSON result from "%s"' % outb)
+
+    def put_device_metrics(self, ioctx, devid, data):
+        old_key = datetime.now() - timedelta(
+            seconds=int(self.retention_period))
+        prune = old_key.strftime(TIME_FORMAT)
+        self.log.debug('put_device_metrics device %s prune %s' %
+                       (devid, prune))
+        erase = []
+        with rados.ReadOpCtx() as op:
+            iter, ret = ioctx.get_omap_keys(op, "", 500) # fixme
+            assert ret == 0
+            ioctx.operate_read_op(op, devid)
+            for key, _ in list(iter):
+                if key >= prune:
+                    break
+                erase.append(key)
+        key = datetime.now().strftime(TIME_FORMAT)
+        self.log.debug('put_device_metrics device %s key %s = %s, erase %s' %
+                       (devid, key, data, erase))
+        with rados.WriteOpCtx() as op:
+            ioctx.set_omap(op, (key,), (str(json.dumps(data)),))
+            if len(erase):
+                ioctx.remove_omap_keys(op, tuple(erase))
+            ioctx.operate_write_op(op, devid)
+
+    def show_device_metrics(self, devid, sample):
```

Comment:
In the typical mode where scraping is happening periodically, it would be nice to keep the latest values in memory, so that calls show_device_metrics were usually fast.  Callers (like the GUI) would have to allow for the possibility of blocking on IO, but they would be able to call repeatedly without issuing a gratuitous number of IO to RADOS.